update the default signing algorithm to ed25519 (breaking change)

ckeshava opened the pull request:
## High Level Overview of Change

If no signing algorithm is specified, use ed25519. At the moment, the code uses secp256k1. This causes incoherence between the `fromSeed` and `generateSeed` functions because they use different defaults.

Note: This change causes breakage in downstream classes like Account

<!--
Please include a summary/list of the changes.
If too broad, please consider splitting into multiple PRs.
If a relevant Asana task, please link it here.
-->

### Context of Change

<!--
Please include the context of a change.
If a bug fix, when was the bug introduced? What was the behavior?
If a new feature, why was this architecture chosen? What were the alternatives?
If a refactor, how is this better than the previous implementation?

If there is a design document for this feature, please link it here.
-->

### Type of Change

<!--
Please check relevant options, delete irrelevant ones.
-->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [x] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

### Did you update HISTORY.md?

- [x] Yes
- [ ] No, this change does not impact library users

## Test Plan

<!--
Please describe the tests that you ran to verify your changes and provide instructions so that others can reproduce.
-->
The current unit tests have been updated to use new values.

<!--
## Future Tasks
For future tasks related to PR.
-->

ckeshava committed:
update the default signing algorithm to ed25519. This causes breakage in downstream classes like Account

ckeshava committed:
update HISTORY file

ckniffen commented:
You should add a way to supply algorithm in secret numbers.  Also this should be a major release for some of the packages.

ckeshava commented:
@ckniffen  Are you suggesting that I add an `algorithm` data member to the Account class? I don't see any other relevant classes in the secret-numbers package.

I agree that this change needs to wait for a major revision of the packages.

ckeshava committed:
include trouble shooting steps: broken dependencies

mvadari commented:
Does this count as a breaking change?

ckeshava commented:
The default signing algorithm of rippled's `wallet_propose` method is `secp256k1` (Docs: https://xrpl.org/docs/references/http-websocket-apis/admin-api-methods/key-generation-methods/wallet_propose)

Why do we have this discrepancy (the defaults for JS client library is `ed25519`) ? 

I saw a note about the inability of rippled command line to generate `ed25519` keys, but I don't believe that would have influenced the choice of defaults

ckeshava committed:
update Wallet.fromMnemonic and Wallet.fromRFC1751Mnemonic to use ed25519 as the default signing algorithm

ckeshava committed:
- add unit tests for Wallet.generateSeed() with secp256k1 signing algorithm
- explicitly state the algorithm used with generateSeed(), instead of relying on the defaults

ckeshava committed:
include algorithm data member in the secret-numbers/Account class

ckeshava commented on packages/secret-numbers/src/schema/Account.ts:
```diff
@@ -7,6 +7,9 @@ import {
 } from '../utils'
 
+// TODO: preferably import this type from ripple-keypairs
+// import type { Algorithm } from 'ripple-keypairs'
```
```
Module '"ripple-keypairs"' declares 'Algorithm' locally, but it is not exported. ts(2459)
index.d.ts(2, 15): 'Algorithm' is declared here.
```
I tried to import the `Algorithm` type from `ripple-keypairs`, but I get the above error. 
I'm confused, `Algorithm` is exported here: https://github.com/XRPLF/xrpl.js/blob/be732a4a6b7bdf58d0efff084b76a66f3f625fe6/packages/ripple-keypairs/src/types.ts#L2

ckeshava commented:
@ckniffen  I have added the algorithm data member in this commit: [a234fa9](https://github.com/XRPLF/xrpl.js/pull/2658/commits/a234fa9cc9a03231f9b84651e805eebcc8365eba)

Please review at your convenience

ckeshava committed:
fix lint error

ckeshava committed:
import Algorithm type from ripple-keypairs module

justinr1234 commented:
@ckeshava please fix lint errors and rebase

mvadari commented:
If this is a breaking change, it needs to wait for a major version bump.

ckeshava committed:
Merge branch 'main' into defaultSigningAlg

ckeshava committed:
fic lint errors: use an import alias

ckeshava committed:
Merge branch 'main' into defaultSigningAlg

khancode commented:
> If this is a breaking change, it needs to wait for a major version bump.

We could release this with rippled v2 client library support as that will also require a major version bump.

ckniffen commented on packages/ripple-keypairs/src/index.ts:
```diff
@@ -38,7 +38,7 @@ function generateSeed(
     ? options.entropy.slice(0, 16)
     : randomBytes(16)
-  const type = options.algorithm === 'ed25519' ? 'ed25519' : 'secp256k1'
+  const type = options.algorithm === 'ecdsa-secp256k1' ? 'secp256k1' : 'ed25519'
```
Should this be tested against the enum?

ckeshava replied:
I've lost some of the context here. Which enum are you referring to?
In line 24, I'm using `secp256k1` imported from `'./signing-schemes/secp256k1'`

Are you referring to that import?

justinr1234 committed:
Merge branch 'main' into defaultSigningAlg

justinr1234 commented:
@coderabbit full review

justinr1234 commented:
@coderabbitai full review

ckeshava committed:
Merge branch 'main' into defaultSigningAlg

ckeshava commented:
@achowdhry-ripple  can you please review this PR at your convinience?

mvadari commented:
Note that making too many major bumps in succession (due to breaking changes) contributes to poor DX

ckeshava committed:
Merge branch 'main' into defaultSigningAlg

ckeshava committed:
revert unrelated changes to CONTRIBUTING guidelines

ckeshava committed:
minor updates to test files and CL

ckeshava committed:
revert updates to package-lock file

ckeshava commented on packages/xrpl/src/Wallet/index.ts:
```diff
@@ -229,7 +228,7 @@ export class Wallet {
         masterAddress: opts.masterAddress,
-        algorithm: opts.algorithm,
+        algorithm: opts.algorithm ?? DEFAULT_ALGORITHM,
       })
```
the `fromMnemonic` method has been deprecated for 3 years. It is expected to be removed in the future.
The previous behavior can still be retained by explicitly specifying the `secp256` signing algorithm.

ckeshava committed:
revert package-lock and CONTRIBUTING ifiles to avoid unnecessary changes

ckeshava committed:
Merge branch 'main' into defaultSigningAlg

kuan121 commented:
[> You should add a way to supply algorithm in secret numbers. Also this should be a major release for some of the packages.](https://github.com/XRPLF/xrpl.js/pull/2658#issuecomment-1986532450)

@ckeshava Does this raise any concerns for you, and has this been addressed in the PR?

kuan121 commented on packages/ripple-keypairs/HISTORY.md:
```diff
@@ -2,6 +2,12 @@
 ## Unreleased
 
+### BREAKING CHANGES
+- Update the default signing algorithm in `generateSeed` function to `ed25519`. This brings compatibility with the `fromSeed` function. Users can retrieve the cryptographic material used in the prior versions of the package by explicitly specifying `ecdsa-secp256k1` in the `generateSeed` function parameter.
```
Do we want to be explicit and mention `fromSeed` function of which package?

ckeshava replied:
ok. I updated the message in 987769c9e. How does that look?

kuan121 commented on packages/secret-numbers/src/schema/Account.ts:
```diff
@@ -1,4 +1,7 @@
 import { deriveAddress, deriveKeypair, generateSeed } from 'ripple-keypairs'
+// Use an import alias to avoid name-conflict with the Algorithm type
+// defined in extensions/node_modules/typescript/lib/lib.dom.d.ts
+import type { Algorithm as _Algorithm } from 'ripple-keypairs'
```
Is it a convention to prefix `_` to handle name conflict? Can we call it `SignAlgorithm`?

ckeshava replied:
ok, updated this in 987769c9e

kuan121 commented on packages/xrpl/src/Wallet/index.ts:
```diff
@@ -269,11 +268,10 @@ export class Wallet {
-    if (opts.algorithm === ECDSA.ed25519) {
-      encodeAlgorithm = 'ed25519'
-    } else {
-      // Defaults to secp256k1 since that's the default for `wallet_propose`
```
> Defaults to secp256k1 since that's the default for `wallet_propose`

Is this still accurate? I’m concerned it could lead to confusion if the server and client library defaults diverge.

ckeshava replied:
The `wallet_propose` RPC method in rippled still defaults to `secp256` algorithm. However, this is not a server-client issue.
<img width="816" height="490" alt="image" src="https://github.com/user-attachments/assets/452ffe6a-6911-4d98-9131-82726756417b" />


The client library is not making any calls to the `wallet_propose` RPC method. Rather, this is a standalone implementation of the Wallet-cryptographic material generation, which is different from the rippled C++ implementation of `wallet_propose`.

I agree with you that it would have been nice to have the same "defaults" in all the Wallet-related RPCs of XRPL ecosystem. However, this would involve a major feature update because rippled-CLI does not even support ED25519 algorithm at the moment.

<img width="803" height="112" alt="image" src="https://github.com/user-attachments/assets/c26dd2c9-374f-4803-99ad-ee3fd11f5714" />


Documentation for `wallet_propose`: https://xrpl.org/docs/references/http-websocket-apis/admin-api-methods/key-generation-methods/wallet_propose

ckeshava committed:
address PR comments

ckeshava committed:
Merge branch 'main' into defaultSigningAlg

ckeshava commented:
> [> You should add a way to supply algorithm in secret numbers. Also this should be a major release for some of the packages.](https://github.com/XRPLF/xrpl.js/pull/2658#issuecomment-1986532450)
> 
> @ckeshava Does this raise any concerns for you, and has this been addressed in the PR?

@kuan121  thanks for drawing attention to this matter.
Yes, this concern has been addressed in the diff-changes in the below displayed file. The `Account` class in the `secret-numbers` package now accepts an `algorithm` value in its constructor. This lets users specify the cryptographic-signing-algorithm of their choice.

<img width="1523" height="773" alt="image" src="https://github.com/user-attachments/assets/111c7be0-5cd9-4596-80c9-bcb665cfb619" />

Furthermore, these unit tests specifically test the migration path for users who created `Account` using the older version of the library. i.e. users can explicitly specify `secp256` to retrieve the previously generated cryptographic material, as shown in the tests. 

<img width="1735" height="830" alt="image" src="https://github.com/user-attachments/assets/0a8a42cd-d1d7-4a50-90c4-b718360ab8d1" />


I also agree that this is a BREAKING CHANGE, as indicated by the Changelog messages and the PR title. However, I did not update the version here because we update the versions in `release`-candidate branches.

ckeshava committed:
Merge branch 'main' into defaultSigningAlg

ckeshava committed:
Merge branch 'main' into defaultSigningAlg

ckeshava commented:
Due to the difficulty of signing historical commits, this work was cherry-picked and committed from a different PR: https://github.com/XRPLF/xrpl.js/pull/3273

Thank you for the reviews.
